pkg/metering/writer.go

```diff
@@ -0,0 +1,191 @@
+	meteringwriter "github.com/pingcap/metering_sdk/writer/metering"
+
+	"github.com/tikv/pd/pkg/utils/logutil"
+)
+
+// meteringFlushInterval is the interval to flush the metering to the underlying storage.
+const meteringFlushInterval = time.Minute
+
+// Collector collects events into records with caller-defined fields.
+type Collector interface {
+	Category() string
+	Collect(data any)
+	Flush() []map[string]any
+}
+
+// Writer is used to as a delegate to collect and flush the metering data to the underlying storage.
+type Writer struct {
+	sync.RWMutex
+	id       string
+	ctx      context.Context
+	cancel   context.CancelFunc
+	wg       sync.WaitGroup
+	provider storage.ObjectStorageProvider
+	inner    *meteringwriter.MeteringWriter
+	// collectors holds category -> collector, allowing callers to customize record fields per category.
+	collectors map[string]Collector
+}
+
+// NewWriter creates a new metering writer to collect and report the metering data to the underlying storage.
+func NewWriter(ctx context.Context, meteringConfig *Config, id string) (*Writer, error) {
+	err := meteringConfig.adjust()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create storage provider
+	providerConfig := &storage.ProviderConfig{
+		Type:   meteringConfig.Type,
+		Bucket: meteringConfig.Bucket,
+		Region: meteringConfig.Region,
+		Prefix: meteringConfig.Prefix,
+	}
+	provider, err := storage.NewObjectStorageProvider(providerConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create storage provider for metering")
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	return &Writer{
+		id:       id,
+		ctx:      ctx,
+		cancel:   cancel,
+		provider: provider,
+		inner: meteringwriter.NewMeteringWriter(
+			provider,
+			config.DefaultConfig().WithLogger(zap.L()),
+		),
+		collectors: make(map[string]Collector),
+	}, nil
+}
+
+// Start starts the metering writer background loop
+func (mw *Writer) Start() {
+	if mw == nil {
+		return
+	}
+	mw.wg.Add(1)
+	go mw.meteringLoop()
+	log.Info("metering writer started")
+}
+
+// Stop stops the metering writer
+func (mw *Writer) Stop() {
+	if mw == nil {
+		return
+	}
+	if mw.cancel != nil {
+		mw.cancel()
+	}
+	mw.wg.Wait()
+	if mw.inner != nil {
+		mw.inner.Close()
+	}
+	log.Info("metering writer stopped")
+}
+
+// RegisterCollector registers a collector for a category. Overwrites existing if any.
+func (mw *Writer) RegisterCollector(collector Collector) {
+	if mw == nil {
+		return
+	}
+	mw.Lock()
+	defer mw.Unlock()
+	if mw.collectors == nil {
+		mw.collectors = make(map[string]Collector)
+	}
+	mw.collectors[collector.Category()] = collector
+}
+
+// meteringLoop runs the background loop to flush metering data periodically.
+func (mw *Writer) meteringLoop() {
+	defer logutil.LogPanic()
+	defer mw.wg.Done()
+
+	ticker := time.NewTicker(meteringFlushInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-mw.ctx.Done():
+			log.Info("metering writer loop exits")
+			return
+		case <-ticker.C:
```

Comment:
I'm afraid the time is not accurate enough in extreme situations.
E.g. The ticker starts at 00:00:59:999.
In the first round, getting the collector is blocked a little and `time.Now()` gets 00:02:00:000, the written minute is 00:02:00.
In the second round, it's fast enough and `time.Now()` gets 00:02:59:999, the written minute is also 00:02:00.
The second round overwrites the first found and the metering data is missing for 1 minute.